packages/shared_preferences/shared_preferences_android/lib/src/shared_preferences_async_android.dart

```diff
@@ -189,10 +207,26 @@ base class SharedPreferencesAsyncAndroid
     final SharedPreferencesPigeonOptions pigeonOptions =
         _convertOptionsToPigeonOptions(options);
     final SharedPreferencesAsyncApi api = _getApiForBackend(pigeonOptions);
-    // TODO(tarrinneal): Remove cast once https://github.com/flutter/flutter/issues/97848
-    // is fixed. In practice, the values will never be null, and the native implementation assumes that.
-    return _convertKnownExceptions<List<String>>(() async =>
-        (await api.getStringList(key, pigeonOptions))?.cast<String>().toList());
+    final Object? dynamicStringList = await _convertKnownExceptions<dynamic>(
+        () async => api.getStringList(key, pigeonOptions));
+    switch (dynamicStringList.runtimeType) {
+      case const (String):
+        final String jsonEncodedString =
+            (dynamicStringList! as String).substring(jsonListPrefix.length);
+        try {
+          final List<String> decodedList =
+              (jsonDecode(jsonEncodedString) as List<dynamic>).cast<String>();
+          return decodedList;
+        } catch (e) {
+          throw TypeError();
+        }
+      case const (List<Object?>):
+        return (dynamicStringList! as List<Object?>).cast<String>().toList();
+      case const (Null):
+        return null;
+      default:
+        throw TypeError();
+    }
```

Comment:
You can use the new pattern system to avoid the need for force unwraps and casts here:

```dart
    switch (dynamicStringList) {
      case String s:
        final String jsonEncodedString =
            s.substring(jsonListPrefix.length);
        try {
          final List<String> decodedList =
              (jsonDecode(jsonEncodedString) as List<dynamic>).cast<String>();
          return decodedList;
        } catch (e) {
          throw TypeError();
        }
      case List<Object?> l:
        return l.cast<String>().toList();
      case null:
        return null;
      default:
        throw TypeError();
    }
```